src/batch.js

```diff
@@ -2,27 +2,243 @@
 
 const { HttpError } = require('./httpError')
 
+const methods = ['GET', 'POST', 'PUT', 'PATCH', 'DELETE']
 const batchDefaultConfig = {
   maxBatchSize: 20
 }
-
-module.exports = {
-  batchHandler
+const batch = {
+  batchHandler,
+  executeRequest,
+  validateBatchRequest,
+  validateDependencyChain,
+  composeQueryStringParametersFromUrl
 }
+module.exports = batch
 
-async function batchHandler({ routes, config, extractPathParameters }, event, context) {
+async function batchHandler({ route, config }, event, context) {
   config = { ...batchDefaultConfig, ...config }
+  const body = event.body
 
   // Validate
-  // validate json body has been parsed
-  // validate body.requests has array of requests
-  // validate each body.requests[]
+  batch.validateBatchRequest(body, config.maxBatchSize)
+
+  //Traverse dependency tree and execute requests
+  let remainingRequests = [...body.requests]
+  const idToPromise = {}
+  do {
+    let toExecute
+    ;[toExecute, remainingRequests] = remainingRequests.reduce(
+      (arr, req) => {
+        if (
+          !req.dependsOn ||
+          req.dependsOn.every(dependency => Object.keys(idToPromise).includes(dependency))
+        ) {
+          arr[0].push(req)
+        } else {
+          arr[1].push(req)
+        }
+        return arr
+      },
+      [[], []]
+    )
+
+    for (const req of toExecute) {
+      idToPromise[req.id] = new Promise(resolve => {
+        ;(async () => {
+          if (req.dependsOn) {
+            for (const id of req.dependsOn) {
+              await idToPromise[id]
+            }
+          }
+
+          resolve(
+            await batch.executeRequest(route, event, context, req).catch(err => {
+              return {
+                response: {
+                  statusCode: 500,
```

Comment:
This should follow the main router's [error handling](https://github.com/Nike-Inc/lambda-router/blob/master/src/index.js#L142-L152) for status code and body

Reply:
Does this look fine or would you prefer it to be a method in the router.js and passed in to batch instead:
https://github.com/Nike-Inc/lambda-router/pull/7/files#diff-13f1b4d1b3c093b14baf4f0e8c9e1558617a85ab30f042d81351d768aa59b612R54

Reply:
I think it should be passed in for two reasons.

1. To reduce duplication
1. The config its using is the batch config, but `includeErrorStack` should be coming from the router's config. The batch handler isn't getting that currently.

Reply:
After moving this around I realized that maybe we don't even need this catch or error handling in batch.js?

This catch:
https://github.com/Nike-Inc/lambda-router/blob/master/src/index.js#L142-L152

Already catches any errors thrown by the client, the only thing the .catch() would do is catch something thrown in the router.route method before the middleware stuff.


Removing it might be for the best as having the private variable onErrorFormat propagate was going to be messy.

Reply:
That catch wont get hit because the error is getting caught here and not re-thrown. The formatting is also important, because the error response for an individual route should match the error response in  `responses[id].body`.

Reply:
I must be misunderstanding.

At the moment, I have a change to remove the .catch(..) from here:
https://github.com/Nike-Inc/lambda-router/pull/7/files#diff-13f1b4d1b3c093b14baf4f0e8c9e1558617a85ab30f042d81351d768aa59b612R55

The resulting output on uncaught exception in a handler is:

```
"body": [
        {
            "id": "A",
            "status": 500,
            "body": {
                "message": "uncaught Error!!!",
                "stack": "Error: uncaught Error!!!\n    at Object.handler ... "
            },
            "headers": {
                "Content-Type": "application/json",
                "Access-Control-Allow-Origin": "*",
                "X-Correlation-Id": "7479600b-8bcd-1067-85ad-bd0a68e0f87b"
            }
        }
        ...
```

Are you saying we should still have that catch()?